DnsTube.Core/Services/IpAddressService.cs

```diff
@@ -30,8 +31,51 @@ public IpAddressService(ISettingsService settingsService, ILogService logService
 
 			var settings = await _settingsService.GetAsync();
 			var url = protocol == IpSupport.IPv4 ? settings.IPv4_API : settings.IPv6_API;
-			var httpClientName = protocol == IpSupport.IPv4 ? HttpClientName.IpAddressV4 : HttpClientName.IpAddressV6;
-			var httpClient = _httpClientFactory.CreateClient(httpClientName.ToString());
+
+			HttpClient httpClient;
+
+			if (protocol == IpSupport.IPv4){
+				httpClient = _httpClientFactory.CreateClient(HttpClientName.IpAddressV4.ToString());
+			}
+			else
+			{
+				httpClient = new HttpClient(new SocketsHttpHandler()
```

Comment:
Creating a new `HttpClient` each time can cause resource starvation issues. Can you do this using `HttpClientFactory`?

Reference: https://learn.microsoft.com/en-us/dotnet/fundamentals/runtime-libraries/system-net-http-httpclient#instancing

Reply:
Using IHttpClientFactory.CreateClient(String) creates a new client every time [[1]](https://learn.microsoft.com/en-us/dotnet/api/system.net.http.ihttpclientfactory.createclient?view=dotnet-plat-ext-8.0). Simply using it, does not instantiate a singleton. We would need to setup singletons for each type of configurations to achieve a single instantiation for each protocol. What I did is not different from what was already being done. I agree it is not elegant, but it achieves the same result: a client is created in every call to GetPublicIpAddressAsync.

I don't see a problem with that, since the client can be disposed after the method finishes and the function is called a maximum of two times every minute, depending on the configuration at the settings page. It may cause socket starvation if called many times and not disposed immediately, which is not the case. On the first paragraph of the reference you linked:

> If you instantiate an HttpClient class for every request, the number of sockets available under heavy loads will be exhausted. This exhaustion will result in [SocketException](https://learn.microsoft.com/en-us/dotnet/api/system.net.sockets.socketexception) errors.

Stress on under heavy loads.

The example of your reference also instantiates the client on the constructor and keep it alive for the duration of the app.

Perhaps we should move both client instantiations to the constructor and use them according to the protocol type.

Reply:
I could be wrong, but I don't think the behavior is the same at all. My understanding is that when using `IHttpClientFactory` it is not required to create a singleton, and that creating a new `HttpClient` each time is recommended practice. This is because it is the underlying `HttpMessageHandler` that is pooled, and does not go away when you dispose an `HttpClient` and later create a new one, so the need for a singleton is greatly reduced. This is what prevents connections from accumulating needlessly.

See https://learn.microsoft.com/en-us/dotnet/core/extensions/httpclient-factory#httpclient-lifetime-management

My take away from that is:

- Cache the `HttpClientHandler` instances
- Create and dispose `HttpClient` instances with the handler as often as you want
- The underlying HttpMessageHandler instances are pooled for you to mitigate socket exhaustion